academy/4-my-own-chain/server-side.md

```diff
@@ -0,0 +1,1046 @@
+            .filter((event: StringEvent) =>
+                event.attributes.some(
+                    (attribute: Attribute) =>
+                        attribute.key === "module" && attribute.value === "checkers",
+                ),
+            )
+        let eventIndex = 0
+        while (eventIndex < myEvents.length) {
+            // Function yet to be declared
+            await handleEvent(myEvents[eventIndex])
+            eventIndex++
+        }
+    } catch (e) {
+        // Skipping if the handling failed. Most likely the transaction failed.
+    }
+}
+```
+
+This needs a new import:
+
+```typescript [https://github.com/cosmos/academy-checkers-ui/blob/b030b2f/src/server/indexer.ts#L4]
+import { ABCIMessageLog, Attribute, StringEvent } from "cosmjs-types/cosmos/base/abci/v1beta1/abci"
+```
+
+* `while() {}` simplifies the syntax of `await`ing multiple times.
+* The `handleEvents` function only keeps events that have a `message` type.
+* It only keeps events that emanate from the `checkers` module.
+* It uses a new function, `handleEvent`. Create one and put `console.log(event)` inside to explore what this object is and consider what actions you can take with it.
+
+## Handle one event
+
+Define `handleEvent` as follows:
+
+```typescript [https://github.com/cosmos/academy-checkers-ui/blob/b030b2f/src/server/indexer.ts#L120-L134]
+const handleEvent = async (event: StringEvent): Promise<void> => {
+    const isActionOf = (actionValue: string): boolean =>
+        event.attributes.some(
+            (attribute: Attribute) => attribute.key === "action" && attribute.value === actionValue,
+        )
+    if (isActionOf("NewGameCreated")) {
+        // Function yet to be declared
+        await handleEventCreate(event)
+    }
+    if (isActionOf("GameRejected")) {
+        // Function yet to be declared
+        await handleEventReject(event)
+    }
+    if (isActionOf("MovePlayed")) {
+        // Function yet to be declared
+        await handleEventPlay(event)
+    }
+}
+```
+
+* [`NewGameCreated`](https://github.com/cosmos/b9-checkers-academy-draft/blob/9a22cd21/x/checkers/types/keys.go#L50), [`GameRejected`](https://github.com/cosmos/b9-checkers-academy-draft/blob/9a22cd21/x/checkers/types/keys.go#L60), and [`MovePlayed`](https://github.com/cosmos/b9-checkers-academy-draft/blob/9a22cd21/x/checkers/types/keys.go#L66) are constant values defined in your Go code. They were associated with a `key` of [`"action"`](https://github.com/cosmos/b9-checkers-academy-draft/blob/9a22cd21/x/checkers/keeper/msg_server_create_game.go#L52).
+* Because events are arrays of key/value pairs you must go through them to find what you want, unless you proactively index events as part of a later optimization.
+* `handleEvent` uses three new functions: `handleEventCreate`, `handleEventReject`, and `handleEventPlay`. Create them and put `console.log(event)` inside each to explore what these objects are and consider what actions you would take with them.
```

Comment:
In post-code notes, I have generally replaced the first generic "it" with what I _think_ is the relevant code product, in this case "handleEvent". However, it may be worth confirming that I have identified the correct products.

Reply:
From what I have seen it seemed correct.